ambry-cloud/src/main/java/com/github/ambry/cloud/azure/CosmosDataAccessor.java

```diff
@@ -136,16 +164,83 @@ void testConnectivity() {
+   * @return {@code true} if the record was deleted, {@code false} if it was not found.
    * @throws DocumentClientException if the operation failed.
    */
-  ResourceResponse<Document> deleteMetadata(CloudBlobMetadata blobMetadata) throws DocumentClientException {
+  boolean deleteMetadata(CloudBlobMetadata blobMetadata) throws DocumentClientException {
     String docLink = getDocumentLink(blobMetadata.getId());
     RequestOptions options = getRequestOptions(blobMetadata.getPartitionId());
-    return executeCosmosAction(() -> asyncDocumentClient.deleteDocument(docLink, options).toBlocking().single(),
-        azureMetrics.documentDeleteTime);
+    try {
+      // Note: not timing here since bulk deletions are timed.
+      executeCosmosAction(() -> asyncDocumentClient.deleteDocument(docLink, options).toBlocking().single(), null);
+      return true;
+    } catch (DocumentClientException dex) {
+      if (dex.getStatusCode() == HttpURLConnection.HTTP_NOT_FOUND) {
+        // Can happen on retry
+        logger.debug("Could not find metadata for blob {} to delete", blobMetadata.getId());
+        return false;
+      } else {
+        throw dex;
+      }
+    }
+  }
+
+  /**
+   * Delete the blob metadata documents in the CosmosDB collection.
+   * @param blobMetadataList the list of blob metadata documents to delete.
+   * @throws DocumentClientException if the operation failed.
+   */
+  void deleteMetadata(List<CloudBlobMetadata> blobMetadataList) throws DocumentClientException {
+    if (bulkDeleteEnabled) {
+      for (List<CloudBlobMetadata> batchOfBlobs : Utils.partitionList(blobMetadataList, purgeBatchSize)) {
+        bulkDeleteMetadata(batchOfBlobs);
+      }
+    } else {
+      for (CloudBlobMetadata metadata : blobMetadataList) {
+        deleteMetadata(metadata);
+      }
+    }
+  }
+
+  /**
+   * Delete the blob metadata documents from CosmosDB using the BulkDelete stored procedure.
+   * @param blobMetadataList the list of blob metadata documents to delete.
+   * @throws DocumentClientException if the operation failed.
+   */
+  private void bulkDeleteMetadata(List<CloudBlobMetadata> blobMetadataList) throws DocumentClientException {
+    String partitionPath = blobMetadataList.get(0).getPartitionId();
+    RequestOptions options = getRequestOptions(partitionPath);
+
+    // stored proc link provided in config.  Test for it at startup and use if available.
+    String quotedBlobIds =
+        blobMetadataList.stream().map(metadata -> '"' + metadata.getId() + '"').collect(Collectors.joining(","));
+    String query = String.format(BULK_DELETE_QUERY, quotedBlobIds);
+    String sprocLink = cosmosCollectionLink + BULK_DELETE_SPROC;
+    boolean more = true;
+    int deleteCount = 0;
+    double requestCharge = 0;
+    try {
+      while (more) {
+        StoredProcedureResponse response =
+            asyncDocumentClient.executeStoredProcedure(sprocLink, options, new String[]{query}).toBlocking().single();
```

Comment:
when continuation is true, we just try again with the same list of blob IDs. Is the stored procedure idempotent?

Reply:
Yes.  It's idempotent because the sproc first runs the query to get records matching the IN filter, which will simply be smaller if some records have been deleted on prior executions.  Though I was not able to trigger this case during testing, likely because the batch size was small.